### PR TITLE
docs: Add codestyle

### LINKED ISSUE #116: Define a code style file

### PR DESCRIPTION
**Pull request by rodrigosiqueira**
Add codestyle documentation to make kw code under control and facilitate
the contribution from the community.

Closes #116

Signed-off-by: Rodrigo Siqueira <rodrigosiqueiramelo@gmail.com>

### CODE REVIEW

**Comment by rodrigosiqueira** [comment]
Hi @matheustavares 

I added the first draft of kw codestyle documentation with the goal that we start to discuss the code standard that we want for kw. Please, few free for starting any discussion here.

File: documentation/content/codingstyle.rst [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  exceptions.
+
+
+Breaking long lines and strings
+-------------------------------
+
+The default limit on the length of lines is 80 columns and this is a strongly
+preferred limit. This is not a rule to be blindly followed, we understand that
+in some cases we need more columns; however, try to do you best for keeping the
+code under 80 characters.
```
[the remark this answers is not in the document]
Comment (rodrigosiqueira) [reply documentation/content/codingstyle.rst]: I agree with that, but I prefer this:
```
my_long_string="kw ran into an unrecoverable error while trying to parse your file."
my_long_string+=" Do you wish to continue anyway [Y/n]?"
```
What do you think?

Do you agree with this:

> Sometimes long strings are a bit cumbersome to keep under 80 columns, in `kw` we adopt string concatenation for this case as the example below illustrates:
```
my_long_string="kw ran into an unrecoverable error while trying to parse your file."
my_long_string+=" Do you wish to continue anyway [Y/n]?"
```

File: documentation/content/codingstyle.rst [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+Placing Optional Buildins and Spaces
+------------------------------------
+
+Some of the shell buildins may accept or not the reserved word `then` or `do`
+which can be added at the end of the expression or in the next line. In kw we
+put the `then` statement at the end of the expression (after the semicolon),
+and put the closing statement first as the example below illustrates::
```
[the remark this answers is not in the document]
Comment (rodrigosiqueira) [reply documentation/content/codingstyle.rst]: Usually, when we use a condition or a loop reserved word we have to start it and close it. For example, in the bash, if we open an `if` condition we have to close it as `fi`... Anyway, I read this part I realized that we can remove it, in the end, we can just need:

> [..] In kw we put the `then` statement at the end of the expression (after the semicolon), as the example below illustrates::

Do you agree with this change?

File: documentation/content/codingstyle.rst [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+  if [[ EXPRESSION  ]] ; then
```
[the remark this answers is not in the document]
Comment (rodrigosiqueira) [reply documentation/content/codingstyle.rst]: Forget about the number of occurrences of `]] ; then` and `]]; then`, in your opinion which one is the best? Think about the code readability...
I'm a slightly inclined for `]] ; then`, but I don't have strong feelings here.

File: documentation/content/codingstyle.rst [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    {
+      [..]
+      if [ "$?" != 0 ] ; then
+        complain "Did you check if your VM is running?"
+        return 125 # ECANCELED
+      fi
+      [..]
+    }
+
+As you can notice from the examples, we use adopt snake case for function definition, this is valid for all the kw code.
```
[the remark this answers is not in the document]
Comment (rodrigosiqueira) [reply documentation/content/codingstyle.rst]: How about if I create a section focused on test with the following description:

> In general, the unit test functions follow the code style rules. We just have a unique case related to the function name: we add `_Test` and the end of each test name. For example:
`function init_kw_Test()`
Additionally, all test files should end with `_test.sh`, if you do not add this suffix your test will be not automatically recognized by `run_tests.sh`.

Is it good for you?

File: documentation/content/codingstyle.rst [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+Overview
+--------
+
+This is a short document describing the kw preferred coding style.
+
+Indentation
+-----------
+
+We adopt two characters indentations.
```
[the remark this answers is not in the document]
Comment (rodrigosiqueira) [reply documentation/content/codingstyle.rst]: agree

File: documentation/content/codingstyle.rst [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+Avoid tricky expression
+-----------------------
+
+We try to keep things as simple as possible in kw, for this reason, we try to
+**avoid**:
+
+* Multiple assignments on a single line;
+* Function call inside if conditions;
```
[the remark this answers is not in the document]
Comment (rodrigosiqueira) [reply documentation/content/codingstyle.rst]: I agree with you, but some times I found myself trying to test code that uses a function inside `if` conditions which make the test task a little bit hard. Anyway... I have no strong feelings about this as well, however, independently of what we decide we have to create an issue for converting all of this thing.

File: documentation/content/codingstyle.rst [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+example::
+
+  function modules_install_to()
+  {
+    [..]
+  }
+
+For the function returning we try to respect the errno codes, for example::
+
+    function vm_modules_install
```
[the remark this answers is not in the document]
Comment (rodrigosiqueira) [reply documentation/content/codingstyle.rst]: Nice catch hehehe

File: documentation/content/codingstyle.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (rodrigosiqueira) [reply documentation/content/codingstyle.rst]: Cool! I added it to a `conclusion` section.

**Comment by rodrigosiqueira** [comment]
Hi @matheustavares , first of all thank you for the feedback. I added my comments inline.

About the get some stuff from git documentation, what do you think of adopt these:

 - Redirection operators should be written with space before, but no
   space after them.  In other words, write 'echo test >"$file"'
   instead of 'echo test> $file' or 'echo test > $file'.  Note that
   even though it is not required by POSIX to double-quote the
   redirection target in a variable (as shown above), our code does so
   because some versions of bash issue a warning without the quotes.

	(incorrect)
	cat hello > world < universe
	echo hello >$world

	(correct)
	cat hello >world <universe
	echo hello >"$world"
 - We prefer $( ... ) for command substitution; unlike ``, it
   properly nests.  It should have been the way Bourne spelled
   it from day one, but unfortunately isn't.
 - If you want to find out if a command is available on the user's
   $PATH, you should use 'type <command>', instead of 'which <command>'.
   The output of 'which' is not machine parsable and its exit code
   is not reliable across platforms.
 - We use Arithmetic Expansion $(( ... )).
 - Inside Arithmetic Expansion, spell shell variables with $ in front
   of them, as some shells do not grok $((x)) while accepting $(($x))
   just fine (e.g. dash older than 0.5.4).

File: documentation/content/codingstyle.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (matheustavares) [reply documentation/content/codingstyle.rst]: Perfect!

File: documentation/content/codingstyle.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (matheustavares) [reply documentation/content/codingstyle.rst]: Ack!

File: documentation/content/codingstyle.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (matheustavares) [reply documentation/content/codingstyle.rst]: Hmm, personally I prefer `]]; then`. I think that extra space is not needed and, for me, the "loose" `;` gets kind of distracting. But I agree with whichever option you decided to go with.

File: documentation/content/codingstyle.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (matheustavares) [reply documentation/content/codingstyle.rst]: Hmm, maybe I haven't thought about that before, how does that impact the testing?

**Comment by matheustavares** [comment]
> Hi @matheustavares , first of all thank you for the feedback. I added my comments inline.
> 
> About the get some stuff from git documentation, what do you think of adopt these:
[...]

I do like them :) Specially the redirection one. I think using something like `>$file <$other` to be very nice for readability.

File: documentation/content/codingstyle.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (rodrigosiqueira) [reply documentation/content/codingstyle.rst]: Agree with you, see the change in the new version

File: documentation/content/codingstyle.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (rodrigosiqueira) [reply documentation/content/codingstyle.rst]: See the new version and let me know if you agree with my changes

File: documentation/content/codingstyle.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (rodrigosiqueira) [reply documentation/content/codingstyle.rst]: I just dropped the above rule, do you agree or do you want to enforce it here?

**Comment by rodrigosiqueira** [comment]
@matheustavares I updated this commit, let me know if you want any new change or if it is fine. I have to admit that I really like the final result.
Thanks for all your feedback.

**Pull request merged**